streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java

```diff
@@ -336,22 +371,34 @@ public V delete(K key) {
 
     @Override
     public KeyValueIterator<K, V> range(K from, K to) {
-        // we need to flush the cache if necessary before returning the iterator
-        if (cache != null)
-            flushCache();
+        lock.readLock().lock();
+        try {
+            validateStoreOpen();
+            // we need to flush the cache if necessary before returning the iterator
+            if (cache != null)
+                flushCache();
 
-        return new RocksDBRangeIterator<>(db.newIterator(), serdes, from, to);
+            return new RocksDBRangeIterator<>(db.newIterator(), serdes, from, to);
```

Comment:
After the iterator is returned, the lock is released, then the stream thread can still close the RocksDB store while iterator is not closed yet, right?

An alternative approach I was thinking, is to keep track of a volatile "RocksObjectCount" for all associated objects, including RocksIterators / WriteBatch / etc, and let the close() call to only proceed when the count becomes 0. Note for WriteBatch since it is only going to be created by the same stream thread it does not actually matter, but just in case it will be created by different threads in the future.